rust/agents/validator/src/submit.rs

```diff
@@ -96,76 +159,73 @@ impl ValidatorSubmitter {
-                            .await?;
-                        if existing.is_some() {
-                            debug!(
-                                index = queued_checkpoint.index,
-                                "Checkpoint already submitted"
-                            );
-                            continue;
-                        }
-
-                        let signed_checkpoint = self.signer.sign(queued_checkpoint).await?;
-                        self.checkpoint_syncer
-                            .write_checkpoint(&signed_checkpoint)
-                            .await?;
-                        debug!(
-                            index = queued_checkpoint.index,
-                            "Signed and submitted checkpoint"
-                        );
-
-                        // small sleep before signing next checkpoint to avoid rate limiting
-                        sleep(Duration::from_millis(100)).await;
-                    }
-
-                    info!(index = checkpoint.index, "Signed all queued checkpoints");
-
-                    self.metrics
-                        .latest_checkpoint_processed
-                        .set(checkpoint.index as i64);
-
-                    // break out of submitter loop if target checkpoint is reached
-                    reached_target = target_checkpoint.is_some();
-                    break;
+            // drain and sign all checkpoints in the queue
+            for queued_checkpoint in checkpoint_queue.drain(..) {
+                let existing = self
+                    .checkpoint_syncer
+                    .fetch_checkpoint(queued_checkpoint.index)
+                    .await?;
+                if existing.is_some() {
+                    debug!(
+                        index = queued_checkpoint.index,
+                        "Checkpoint already submitted"
+                    );
+                    continue;
                 }
+
+                let signed_checkpoint = self.signer.sign(queued_checkpoint).await?;
+                self.checkpoint_syncer
+                    .write_checkpoint(&signed_checkpoint)
+                    .await?;
+                debug!(
+                    index = queued_checkpoint.index,
+                    "Signed and submitted checkpoint"
+                );
+
+                // small sleep before signing next checkpoint to avoid rate limiting
+                sleep(Duration::from_millis(100)).await;
             }
 
-            sleep(self.interval).await;
-        }
+            info!(
+                index = checkpoint.index,
+                "Signed all queued checkpoints until index"
+            );
 
-        // TODO: remove this once validator is tolerant of tasks exiting
-        loop {
-            sleep(Duration::from_secs(u64::MAX)).await;
+            self.metrics
+                .latest_checkpoint_processed
+                .set(checkpoint.index as i64);
```

Comment:
Nit: I still like this condition ahead of the happy path as a "revert reason" instead of a true if/else branching condition